File: backend/apps/core/modules.py
"""Registry of permission-checked modules, shared by the RBAC system and every app's API views.

Adding a new module means adding one entry here — RolePermission rows for every
existing role are then backfilled automatically (see accounts.permissions.sync_role_permissions).
"""

CLIENTS = "clients"
COMPANIES = "companies"
CATALOG = "catalog"
SUPPLIERS = "suppliers"
PROJECTS = "projects"
ORDERS = "orders"
QUOTATIONS = "quotations"
COSTING = "costing"
REPORTS = "reports"
USERS = "users"
SETTINGS = "settings"
NOTIFICATIONS = "notifications"

MODULES = {
    CLIENTS: "Clients",
    COMPANIES: "Companies",
    CATALOG: "Products",
    SUPPLIERS: "Suppliers",
    PROJECTS: "Projects",
    ORDERS: "Orders",
    QUOTATIONS: "Quotations",
    COSTING: "Costing",
    REPORTS: "Reports",
    USERS: "Users",
    SETTINGS: "Settings",
    NOTIFICATIONS: "Notifications",
}

MODULE_CHOICES = list(MODULES.items())

VIEW = "view"
ADD = "add"
EDIT = "edit"
DELETE = "delete"
ACTIONS = (VIEW, ADD, EDIT, DELETE)

# Maps HTTP methods to the action they require, for the DRF permission class.
METHOD_ACTION_MAP = {
    "GET": VIEW,
    "HEAD": VIEW,
    "OPTIONS": VIEW,
    "POST": ADD,
    "PUT": EDIT,
    "PATCH": EDIT,
    "DELETE": DELETE,
}

File: frontend/src/lib/modules.ts
/** Mirrors apps/core/modules.py on the backend — the RBAC module keys the
 * permission matrix is keyed by. Keep these two files in sync. */
export const MODULES = {
  CLIENTS: "clients",
  COMPANIES: "companies",
  CATALOG: "catalog",
  SUPPLIERS: "suppliers",
  PROJECTS: "projects",
  ORDERS: "orders",
  QUOTATIONS: "quotations",
  COSTING: "costing",
  REPORTS: "reports",
  USERS: "users",
  SETTINGS: "settings",
  NOTIFICATIONS: "notifications",
} as const;

export type ModuleKey = (typeof MODULES)[keyof typeof MODULES];
export type PermissionAction = "view" | "add" | "edit" | "delete";

/** Display labels for the Roles & Permissions matrix — mirrors the MODULES
 * dict values in apps/core/modules.py, same order. */
export const MODULE_LABELS: Record<ModuleKey, string> = {
  [MODULES.CLIENTS]: "Clients",
  [MODULES.COMPANIES]: "Companies",
  [MODULES.CATALOG]: "Products",
  [MODULES.SUPPLIERS]: "Suppliers",
  [MODULES.PROJECTS]: "Projects",
  [MODULES.ORDERS]: "Orders",
  [MODULES.QUOTATIONS]: "Quotations",
  [MODULES.COSTING]: "Costing",
  [MODULES.REPORTS]: "Reports",
  [MODULES.USERS]: "Users",
  [MODULES.SETTINGS]: "Settings",
  [MODULES.NOTIFICATIONS]: "Notifications",
};

export const PERMISSION_ACTIONS: PermissionAction[] = ["view", "add", "edit", "delete"];
